File: clients/sdk/src/guardrails.ts
import type { PublicKey } from "@solana/web3.js";
import { SLOTS_PER_HOUR } from "./constants.js";
import {
  HourlyLimitExceededError,
  LifetimeLimitExceededError,
  NotWhitelistedError,
  PerTxLimitExceededError,
  VaultFrozenError,
  ZeroAmountError,
} from "./errors.js";
import type { CreditVaultAccount, SpendPolicyAccount } from "./types.js";

// Mirrors `check_and_record_spend` in programs/credit_vault/src/policy.rs.
// This is a best-effort pre-flight: a concurrent spend, a window roll-over, or
// any other on-chain state change between this check and the transaction
// landing can still cause the chain to reject. The SDK surfaces the on-chain
// error in that case — the local check just avoids the network round-trip for
// the obvious failures.
export function guardSpend(args: {
  vault: CreditVaultAccount;
  policy: SpendPolicyAccount;
  service: PublicKey;
  amountUsdc: number;
  currentSlot: number;
}): void {
  const { vault, policy, service, amountUsdc, currentSlot } = args;

  if (vault.frozen) throw new VaultFrozenError();
  if (amountUsdc <= 0) throw new ZeroAmountError();

  if (policy.whitelist.length > 0) {
    const allowed = policy.whitelist.some((pk) => pk.equals(service));
    if (!allowed) throw new NotWhitelistedError(service.toBase58());
  }

  if (policy.per_tx_limit_usdc > 0 && amountUsdc > policy.per_tx_limit_usdc) {
    throw new PerTxLimitExceededError(amountUsdc, policy.per_tx_limit_usdc);
  }

  if (policy.hourly_limit_usdc > 0) {
    const windowElapsed = currentSlot - policy.hourly_window_start_slot >= SLOTS_PER_HOUR;
    const windowSpent = windowElapsed ? 0 : policy.hourly_window_spent_usdc;
    const newWindowTotal = windowSpent + amountUsdc;
    if (newWindowTotal > policy.hourly_limit_usdc) {
      throw new HourlyLimitExceededError(amountUsdc, windowSpent, policy.hourly_limit_usdc);
    }
  }

  if (policy.lifetime_limit_usdc > 0) {
    const newTotalSpent = vault.total_spent + amountUsdc;
    if (newTotalSpent > policy.lifetime_limit_usdc) {
      throw new LifetimeLimitExceededError(amountUsdc, vault.total_spent, policy.lifetime_limit_usdc);
    }
  }
}

File: clients/sdk-python/src/agent_fuel/guardrails.py
"""Local mirror of `check_and_record_spend` from
programs/credit_vault/src/policy.rs. Catches the obvious failures
(frozen, zero, whitelist, per-tx, hourly, lifetime) before we burn a
transaction slot — but the on-chain ladder is authoritative; this is a
pre-flight courtesy."""

from __future__ import annotations

from solders.pubkey import Pubkey

from .constants import SLOTS_PER_HOUR
from .errors import (
    HourlyLimitExceededError,
    LifetimeLimitExceededError,
    NotWhitelistedError,
    PerTxLimitExceededError,
    VaultFrozenError,
    ZeroAmountError,
)
from .types import CreditVaultAccount, SpendPolicyAccount


def guard_spend(
    *,
    vault: CreditVaultAccount,
    policy: SpendPolicyAccount,
    service: Pubkey,
    amount_usdc: int,
    current_slot: int,
) -> None:
    if vault.frozen:
        raise VaultFrozenError()
    if amount_usdc <= 0:
        raise ZeroAmountError()
    if policy.whitelist and service not in policy.whitelist:
        raise NotWhitelistedError(str(service))
    # On-chain semantics (programs/credit_vault/src/policy.rs): a limit of
    # `0` means "no enforcement", not "zero allowed". Each check is gated
    # on `if policy.X_limit_usdc > 0`. Mirror that here or this guardrail
    # rejects every spend on an unlimited policy.
    if policy.per_tx_limit_usdc > 0 and amount_usdc > policy.per_tx_limit_usdc:
        raise PerTxLimitExceededError(amount_usdc, policy.per_tx_limit_usdc)
    # Hourly window slides every SLOTS_PER_HOUR slots. Spent counter resets
    # on chain when a new window opens; mirror that here so a stale value
    # from a closed window doesn't trip a false-positive rejection.
    if policy.hourly_limit_usdc > 0:
        window_spent = policy.hourly_window_spent_usdc
        if current_slot - policy.hourly_window_start_slot >= SLOTS_PER_HOUR:
            window_spent = 0
        if amount_usdc + window_spent > policy.hourly_limit_usdc:
            raise HourlyLimitExceededError(
                amount_usdc, window_spent, policy.hourly_limit_usdc
            )
    if (
        policy.lifetime_limit_usdc > 0
        and amount_usdc + vault.total_spent > policy.lifetime_limit_usdc
    ):
        raise LifetimeLimitExceededError(
            amount_usdc, vault.total_spent, policy.lifetime_limit_usdc
        )
